chapter_recurrent-neural-networks/sequence.ipynb

```diff
@@ -0,0 +1,576 @@
+    "import ai.djl.training.*;\n",
+    "import ai.djl.training.dataset.*;\n",
+    "import ai.djl.training.initializer.*;\n",
+    "import ai.djl.training.loss.*;\n",
+    "import ai.djl.training.optimizer.*;\n",
+    "import ai.djl.training.tracker.*;\n",
+    "import ai.djl.training.listener.*;\n",
+    "import ai.djl.translate.*;\n",
+    "import ai.djl.inference.*;\n",
+    "import ai.djl.Model;"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "NDManager manager = NDManager.newBaseManager();"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public Figure plot(double[] x, double[] y, String xLabel, String yLabel) {\n",
+    "    ScatterTrace trace = ScatterTrace.builder(x,y)\n",
+    "        .mode(ScatterTrace.Mode.LINE)\n",
+    "        .build();\n",
+    "\n",
+    "    Layout layout = Layout.builder()\n",
+    "            .showLegend(true)\n",
+    "            .xAxis(Axis.builder().title(xLabel).build())\n",
+    "            .yAxis(Axis.builder().title(yLabel).build())\n",
+    "            .build();\n",
+    "\n",
+    "    return new Figure(layout, trace);\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "int T = 1000; //Generate a total of 1000 points\n",
+    "NDArray time = manager.arange(1f, T+1);\n",
+    "NDArray x = time.mul(0.01).sin().add(\n",
+    "    manager.randomNormal(0f, 0.2f, new Shape(T), DataType.FLOAT32));\n",
+    "\n",
+    "double[] xAxis = Functions.floatToDoubleArray(time.toFloatArray());\n",
+    "double[] yAxis = Functions.floatToDoubleArray(x.toFloatArray());\n",
+    "\n",
+    "plot(xAxis, yAxis, \"time\", \"x\");"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 6
+   },
+   "source": [
+    "Next, we need to turn such a sequence into features and labels that our model can train on.\n",
+    "Based on the embedding dimension $\\tau$ we map the data into pairs $y_t = x_t$ and $\\mathbf{x}_t = [x_{t-\\tau}, \\ldots, x_{t-1}]$.\n",
+    "The astute reader might have noticed that this gives us $\\tau$ fewer data examples, since we do not have sufficient history for the first $\\tau$ of them.\n",
+    "A simple fix, in particular if the sequence is long,\n",
+    "is to discard those few terms.\n",
+    "Alternatively we could pad the sequence with zeros.\n",
+    "Here we only use the first 600 feature-label pairs for training.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "int tau = 4;\n",
+    "NDArray features = manager.zeros(new Shape(T - tau, tau));\n",
+    "\n",
+    "for (int i = 0; i < tau; i++) {\n",
+    "    features.set(new NDIndex(\":, \" + i), x.get(new NDIndex(\"\" + i + \":\" + (T - tau + i))));\n",
```

Comment:
Instead of using string concatenation, you should be able to use `new NDIndex(":, {}", i)` and `new NDIndex("{}:{}", i, T - tau + i)`. Also applies below

Reply:
Cool, just saw it in the documentation. I wasn't aware so thanks for sharing. Is it normally preferred over string concatenation?

Reply:
I think it is a bit cleaner than the concatenation. It is easier to see how the indices fit together which can get a bit lost in the concatenation

Reply:
I see. I'll change it now then.

Reply:
Done.